src/hotspot/share/oops/compressedKlass.cpp

```diff
@@ -44,24 +44,33 @@ narrowKlass CompressedKlassPointers::_lowest_valid_narrow_klass_id = (narrowKlas
 narrowKlass CompressedKlassPointers::_highest_valid_narrow_klass_id = (narrowKlass)-1;
 size_t CompressedKlassPointers::_protection_zone_size = 0;
 
-#ifdef _LP64
-
 size_t CompressedKlassPointers::max_klass_range_size() {
-  // We disallow klass range sizes larger than 4GB even if the encoding
-  // range would allow for a larger Klass range (e.g. Base=zero, shift=3 -> 32GB).
-  // That is because many CPU-specific compiler decodings do not want the
-  // shifted narrow Klass to spill over into the third quadrant of the 64-bit target
-  // address, e.g. to use a 16-bit move for a simplified base addition.
-  return MIN2(4 * G, max_encoding_range_size());
+#ifdef _LP64
+  const size_t encoding_allows = nth_bit(narrow_klass_pointer_bits() + max_shift());
+  constexpr size_t cap = 4 * G;
+  return MIN2(encoding_allows, cap);
+#else
+  // 32-bit: only 32-bit "narrow" Klass pointers allowed. If we ever support smaller narrow
+  // Klass pointers here, coding needs to be revised.
+  // We keep one page safety zone free to guard against size_t overflows on 32-bit. In practice
+  // this is irrelevant, since these upper address space parts are not user-addressable on
+  // any of our 32-bit platforms.
+  return align_down(UINT_MAX, os::vm_page_size());
+#endif
 }
 
 void CompressedKlassPointers::pre_initialize() {
   if (UseCompactObjectHeaders) {
     _narrow_klass_pointer_bits = narrow_klass_pointer_bits_coh;
     _max_shift = max_shift_coh;
   } else {
+#ifdef _LP64
```

Comment:
It's a bit confusing when we need NEEDS_CLASS_SPACE vs. LP64.  Like here.

Reply:
@coleenp I modified the patch, removed the NEEDS_CLASS_SPACE macro, since most of the sections can be compiled just fine on arm. There is the basic question of whether one wants to exclude anything related to 32-bit on 64-bit from builds or not. If one does, 32-bit will bitrot even easier, if not (this seems e.g. the strategy Ioi is following in most of his CDS code) one sees errors in 32-bit more easily and reduces ifdef sections. I opted for the latter here, but it won't make much difference in practice I assume.